Adds zipWith operation on flows

jmgimeno opened the pull request:
This pull request introduces a new `zipWith` method for combining elements of two flows in the `Async` module and includes corresponding updates to the test suite. The most important changes include the addition of the `zipWith` method, its implementation using semaphores for synchronization, and a new test case to validate its functionality.

### New feature: `zipWith` method

* **`zipWith` method added to `Flow` class**: Introduced a new method `zipWith` in `Flow` to combine elements from two flows using a provided function. This method ensures synchronization using semaphores and emits elements only when both flows have values.

### Supporting changes

* **Semaphore import**: Added `java.util.concurrent.Semaphore` to support the synchronization logic in the `zipWith` implementation.

### Testing updates

* **New test case for `zipWith`**: Added a test case in `AsyncSpec` to verify that the `zipWith` method correctly combines corresponding elements from two flows using a specified function.

rcardin commented on yaes-core/src/test/scala/in/rcard/yaes/AsyncSpec.scala:
```diff
@@ -548,4 +548,16 @@ class AsyncSpec extends AnyFlatSpec with Matchers {
+
+    actualResult should contain theSameElementsInOrderAs Seq("a" -> 1, "b" -> 2, "c" -> 3)
+  }
 }
```
Can you add a test with an empty flow? The result should be an empty flow as well. What if one of the two flows is canceled? Is it handled? We need to consider many edge cases, not only the happy path.

rcardin replied:
You can cancel the fibers used to run it 😄

rcardin commented on yaes-core/src/main/scala/in/rcard/yaes/Async.scala:
```diff
@@ -266,6 +266,80 @@ object Async {
+     */
+    def zipWith[B, C](other: Flow[B])(f: (A, B) => C): Flow[C] =
+      Flow.flow {
+        Async.run {
```
I don't want the library to run the effect in a hidden way. The end user should always call the `Async.run` method. Moreover, I want to try to implement the `zipWith` (or `zip`) using the available structured concurrency primitives. We have to understand if it's possible. 

The Kotlin implementation uses the `Channel<T>` type: https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/flow/internal/Combine.kt#L82

rcardin commented:
@jmgimeno, first of all, thanks for your commitment ❤️.
I made some comments on the PR. In my opinion, we need to understand if it's possible to implement the `zip` function using the available structured concurrency primitives and, if not, which one we need to add to make it possible.

Please let me know what you think.

jmgimeno commented:
I've updated my solution using Async.race but I suspect I have a race condition: when one of the fibesr that are collecting values from one of the input flows ends, that is, when it gets the last value, the other fibers are cancelled, and maybe the emission fiber hasn't emitted the last value yet. 

Am I right suspecting that?

Semaphores have a method acquireUninterruptedly that I don't know if it can help with that suspected race.

rcardin commented:
> I've updated my solution using Async.race but I suspect I have a race condition: when one of the fibesr that are collecting values from one of the input flows ends, that is, when it gets the last value, the other fibers are cancelled, and maybe the emission fiber hasn't emitted the last value yet.
> 
> Am I right suspecting that?

As I said in my last comment, the `race` function cancels the losing fiber. Try to introduce a delay in value emissions.

jmgimeno commented on yaes-core/src/main/scala/in/rcard/yaes/Async.scala:
```diff
@@ -235,37 +235,100 @@ object Async {
+        var b           = null.asInstanceOf[B]
+        Async.race(
+          {
+            Async.race(
```
Ok, I see. I have some ideas that _maybe solve_ this and then I'll consider interruptions in the incoming flows.

I've always hated dealing with this concurrency problems 💔

rcardin replied:
No problem. I want to take some time to understand how it is implemented in the Kotlinx library. As I said, they use communication channels. I will let you know

rcardin replied:
@jmgimeno, I think the `SynchronousQueue` is what you're searching for: https://www.baeldung.com/java-synchronous-queue

jmgimeno replied:
@rcardin But in our case we have two producers and one consumer. 

If we have one queue between each producer and the consumer, we cannot block waiting for both queues at the same time.

rcardin replied:
Let me try to implement my idea. I'll try during the day ☺️

jmgimeno replied:
Interesting idea .... I think I can adapt some of the things I did on my last solution to deal with termination (and ensuring the last element is always treated), along these lines:

```scala
    def zipWith[B, C](other: Flow[B])(f: (A, B) => C)(using async: Async): Flow[C] = Flow.flow {
      val second: SynchronousQueue[Option[B]] = new SynchronousQueue();
      Async.race(
        {
          other.collect { b =>
            second.put(Some(b))
          }
          second.put(None)
        }, {
          flow.collect { a =>
            second.take() match
              case None    => ()
              case Some(b) => Flow.emit(f(a, b))
          }
        }
      )
    }
```
Now, when one of the fibers is finished, the race with interrupt the other. We need the poison pill to ensure that the second one has emitted the element before being interrupted. 

I'll try to

jmgimeno replied:
In order to control the finalization when one of the flows is cancelled, I've changed `Async.race` to `Async.par'  and ensured that both fibers ends with a `Raise.raise`. The code is:

```scala
    def zipWith[B, C](other: Flow[B])(f: (A, B) => C)(using async: Async): Flow[C] = Flow.flow {
      val second: SynchronousQueue[Option[B]] = new SynchronousQueue()
      Raise.run {
        Async.par(
          {
            other.collect { b =>
              second.put(Some(b))
            }
            second.put(None)
            Raise.raise(())
          }, {
            flow.collect { a =>
              second.take() match
                case None    => ()
                case Some(b) => Flow.emit(f(a, b))
            }
            Raise.raise(())
          }
        )
      }
    }
  }
```

But when I run it i get this error:

```
scala.util.boundary$Break was thrown.
scala.util.boundary$Break
```
and no indication about what happens. Any idea?

jmgimeno replied:
I've adapted your solution and I need to study your tests in AsyncSpec to understand how to create the scenarios.

rcardin replied:
Ok, don't hesitate to let me know if anything is unclear.

jmgimeno commented:
@rcardin In these two last commits, I've added more rigorous testing to `zipWith`. To do so, I've added `scalacheck` to the project dependencies and, to isolate these tests from the rest of the testing, I've created a new class for the tests of `zipWith`. 

My next move, time allowing, will be to create a new battery of tests to deal with interruption.

jmgimeno commented:
I'm stuck generating flows that get interrupted in the middle of its generation. I've created a simple test:

```scala
  "A flow" should "stop emitting elements when interrupted" in {
    Async.run {
      val flow = Flow.flow {
        Raise.withDefault(()) {
          Async.timeout(125.millis) {
            Flow.emit('A')
            Async.delay(50.millis)
            Flow.emit('B')
            Async.delay(50.millis)
            Flow.emit('C')
            Async.delay(50.millis)
            Flow.emit('D')
            Async.delay(50.millis)
            Flow.emit('E')
          }
        }
      }
      val queue = ConcurrentLinkedQueue[Char]()
      flow.collect { value =>
        queue.add(value)
      }
      queue.toArray should contain theSameElementsInOrderAs Seq('A', 'B', 'C')
    }
  }
```
that should pass, but its execution only prints:
```
scala.util.boundary$Break was thrown.
scala.util.boundary$Break
```
Any idea of what's going on?

rcardin commented:
First, you should move the assertions outside the `Async.run`.

However, I cannot understand what you're trying to test. A test already verifies that a flow stops if an exception is thrown during its execution.

jmgimeno commented:
This is the way, parameterizing some things and after some dead-ends, that I have to create flows that will be cancelled in the middle. This is needed for the interruption scenarios for the `zipWith` tests.

I've moved the assert out of the `Async.run`

```scala
  "A flow" should "stop emitting elements when interrupted" in {
    val queue = ConcurrentLinkedQueue[Char]()
    Async.run {
      val flow = Flow.flow {
        Raise.withDefault(()) {
          Async.timeout(125.millis) {
            Flow.emit('A')
            Async.delay(50.millis)
            Flow.emit('B')
            Async.delay(50.millis)
            Flow.emit('C')
            Async.delay(50.millis)
            Flow.emit('D')
            Async.delay(50.millis)
            Flow.emit('E')
          }
        }
      }
      flow.collect { value =>
        queue.add(value)
      }
    }
    queue.toArray should contain theSameElementsInOrderAs Seq('A', 'B', 'C')
  }
```
but it fails the same way.

rcardin commented:
Ok, I’ll try to make some tries during the weekend. Sorry for the delay 😅

rcardin commented:
@jmgimeno, I found how to make it work:

```scala 3
"A flow" should "stop emitting elements when interrupted" in {
  val queue = ConcurrentLinkedQueue[Char]()
  Raise.run {
    Async.run {
      Async.timeout(125.millis) {
        val flow: Flow[Char] = Flow.flow[Char] {
          Flow.emit('A')
          Async.delay(50.millis)
          Flow.emit('B')
          Async.delay(50.millis)
          Flow.emit('C')
          Async.delay(50.millis)
          Flow.emit('D')
          Async.delay(50.millis)
          Flow.emit('E')
        }
        flow.collect { value =>
          queue.add(value)
        }
      }
    }
  }
  queue.toArray should contain theSameElementsInOrderAs Seq('A', 'B', 'C')
}
```

However, I don't understand why your solution is not working. You probably found a bug 😅. I need to dig deeper. In the meantime, you can continue using the code I shared.

jmgimeno commented:
Thanks !! I'll try to adapt the idea I have for the test to this structure, butI have a busy week, so I don't know when I'll be abre to return to this.

rcardin commented:
@jmgimeno, I fixed the bug (to be fair, Copilot did it...). By the way, if you merge the `main` into your branch, you'll get the fix and can use your original version of the test.

jmgimeno commented:
With the fix, the single test I had for testing interruption of flows when zipping passes. 

When I have time, I'll transform it to a property-based test to check it with randomized delays.

rcardin commented:
Is it completed?

jmgimeno commented:
@rcardin The tests generate scenarios for:
* zipping flows that emit values with random delays between them
* zipping those flows when a flow gets interrupted in a point in time
So, yes, I think they're complete (if that thing is ever possible to say)
